Zend/zend_objects_API.h

```diff
@@ -81,9 +81,17 @@ static zend_always_inline void zend_object_release(zend_object *obj)
 
 static zend_always_inline size_t zend_object_properties_size(zend_class_entry *ce)
 {
-	return sizeof(zval) *
-		(ce->default_properties_count -
-			((ce->ce_flags & ZEND_ACC_USE_GUARDS) ? 0 : 1));
+	size_t size = ce->default_properties_count - 1;
+
+	if (ce->ce_flags & ZEND_ACC_USE_GUARDS) {
+		size++;
+	}
+
+	if (ce->ce_flags & ZEND_ACC_USE_WEAK_REFS) {
```

Comment:
Won't this make pretty much every object 16 byte larger?

Reply:
We should make sure that weak refs are (essentially) free if they aren't used. Instead of always reserving a slot in each object, I'd instead suggest to add a global HT `EG(weakrefs)` that's a map from object address => HT of weakrefs and only store a 1-bit flag in the object whether the object has an entry in EG(weakrefs).